### PR TITLE
Make inventory UI great again

### LINKED ISSUE #935: Basic Inventory UI
## Summary

Create a basic inventory UI for the inventory system.

- The user has to be able to pick items and see those in their hand slots.

## Goal 

Provide the basis for inventory and see how bad the current container system is.

### LINKED ISSUE #939: Hands and Inventory UI
## Summary

The game currently already has items and interactions, but the problem is that there is no UI related to that, meaning you can't see what items you have picked up, also preventing us from working on features like crafting and item-to-item interactions, you can just read how it was done in the "outdated" branch and replicate it, doing some small adaptations to the new systems.

## Goal 

- [ ] Having a UI for Hands
- [ ] Seeing which items you are currently holding
- [ ] Knowing which hand is the currently active one
- [ ] Optional: UI for the rest of the rest of the inventory(Pockets, Backpack, etc)

### PR DESCRIPTION
**Pull request by stilnat**
## Summary

Adds a functionning UI for the hands, and some placeholder for the ID, Bags, and pockets.
Reintroduce the sprite generator.
Make a few small cosmetic changes on the Inventory UI (rotation, outliner, spacing, size...).

## Pictures/Videos (optional)

image

## Technical Notes (optional)

No major changes. The main change is that Inventory is now activating inventoryUI.

## Known issues

- The hands can't be switched by clicking on the icons like on the outdated branch. This can be solved in a different PR.
- The sprite are kind of ugly, could be due to the shader, maybe that can also be part of another PR.

## Fixes (optional)

Closes #935 
Closes #939

### CODE REVIEW

**Comment by cosmiccoincidence** [comment]
> bring an item from one hand over to the other and that item's ui pretty much breaks and you can't drag anything anymore. - LightVelox

Seems there seems to be some bugs. Is that due to lack of container functionality or a problem with the UI itself?

**Comment by stilnat** [comment]
> > bring an item from one hand over to the other and that item's ui pretty much breaks and you can't drag anything anymore. - LightVelox
> 
> Seems there seems to be some bugs. Is that due to lack of container functionality or a problem with the UI itself?

I'll check that as soon as I can get in game again, I have a bunch of assembly errors right now

**Comment by stilnat** [comment]
@cosmiccoincidence Where did you see LightVelox comment ? I would have missed it if you didn't cite it

**Comment by Ryan089** [comment]
@stilnat it was in the #pull-requests channel [here](https://discord.com/channels/483813139633143808/483814334770839573/1058220429321457766).

**Review by TaylorNAlbarnaz — CHANGES_REQUESTED** [review]
You should only be able to drag items with your Left Mouse Button, currently you can drag using the Right Mouse Button as well

**Comment by stilnat** [comment]
Don't accept this PR yet, it's not working client side, only host side

**Comment by stilnat** [comment]
Nevermind, the bug is only in the client when client is in editor

File: Assets/Scripts/SS3D/Systems/Storage/Containers/AttachedContainer.cs [hunk 1]
```diff
@@ -150,7 +150,9 @@ void handleItemAdded(Item item)
 
             void handleItemRemoved(Item item)
             {
-                item.Unfreeze();
+                // Only unfreeze the item if it was not just placed into another container
+                if(item.Container == null)
+                    item.Unfreeze();
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Storage/Containers/AttachedContainer.cs]: it is preferred to add braces

File: Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs [hunk 2]
```diff
@@ -45,9 +47,13 @@ public class Hands : InteractionSourceNetworkBehaviour, IToolHolder, IInteractio
         /// </summary>
         public bool SelectedHandEmpty => SelectedHandContainer.Empty;
 
+        public HandsUi HandsUI {get; private set;}
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs]: add space before get and after set;

File: Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs [hunk 3]
```diff
@@ -251,6 +252,22 @@ public void SetContainerUnchecked(Container newContainer)
...
         }
 
+        // TODO: Improve this
+        // we have this to generate icons at start, I do not know how bad it is for performance
+        // if you know anything about it, tell us
+        public void GenerateNewIcon()
+        {
+            RuntimePreviewGenerator.BackgroundColor = new Color(0, 0, 0, 0);
+            RuntimePreviewGenerator.OrthographicMode = true;
+
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs]: Remove empty space

File: Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs [hunk 4]
```diff
@@ -251,6 +252,22 @@ public void SetContainerUnchecked(Container newContainer)
...
+            
+            Texture2D texture = RuntimePreviewGenerator.GenerateModelPreviewWithShader(this.transform, Shader.Find("Legacy Shaders/Diffuse"), null, 128, 128, false);
+
+            _sprite = Sprite.Create(texture, new Rect(0, 0, texture.width, texture.height), new Vector2(0.5f, 0.5f), 100);
+
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Storage/Items/Item.cs]: Remove empty space

File: Assets/Scripts/SS3D/Systems/Storage/UI/ItemDisplay.cs [hunk 5]
```diff
@@ -74,6 +74,9 @@ public void OnPointerClick(PointerEventData eventData)
 
         public void OnBeginDrag(PointerEventData eventData)
         {
+            // Only allow to drag with a left click.
+            if(eventData.button != PointerEventData.InputButton.Left) return;
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Storage/UI/ItemDisplay.cs]: add space after if

**Review by joaoburatto — CHANGES_REQUESTED** [review]

**Comment by stilnat** [comment]
The UI is now working on any player, client or host.

**Review by TaylorNAlbarnaz — APPROVED** [review]
The problem was properly fixed and it works as expected now

**Pull request merged**